### PR TITLE
Bugfix/816

### LINKED ISSUE #816: Crashes on Raspberrry Pi when changing LED Layout
- [x] I confirm that this is an issue rather than a question.

## Bug report

Hyperion crashes on Raspberry Pi 4 when trying to change the LED Layout to higher numbers.

#### Steps to reproduce

- install Hyperion on Raspberry Pi 4 via deb
- open WebUI and got to LED Layout settings
- set one of the following Layouts:
    - Top/Bottom: 119, Left/Right: 53
    - Top/Bottom: 12, Left/Right: 5
- click "save"

#### What is expected?

New LED Layout is saved and applied to connected controller.

#### What is actually happening?

Hyperion crashes with following error:

```txt
[1]    14010 segmentation fault  /usr/bin/hyperiond
```

Repeated saves don't do anything. It crashes every time.

Note: with the Windows Version, everything works as expected.

#### System

Hyperion Server: 
- Build:       (HEAD detached at f5583de8) (Sersch-f7865b45/f5583de8-1590611619)
- Build time:  May 27 2020 21:11:33
- Git Remote:  https://github.com/hyperion-project/hyperion.ng
- Version:     2.0.0-alpha.6
- UI Lang:     en (BrowserLang: de-DE)
- UI Access:   expert
- Avail Capt:  dispmanx,v4l2,framebuffer,qt

Hyperion Server OS: 
- Distribution: Raspbian GNU/Linux 10 (buster)
- Arch:         arm
- Kernel:       linux (4.19.118-v7l+ (WS: 32))
- Browser:      Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/83.0.4103.61 Safari/537.36

### PR DESCRIPTION
**Pull request by m-seker**
**Summary**

Fixes #816, maybe #772 too.

Edit : Looks like #609 too.

If the incoming vector is bigger than the member vector, ````memcpy```` overwrites beyond the vector's memory. We should resize() before ````memcpy````, or just use ````operator=````.

**What kind of change does this PR introduce?** (check at least one)

- [x] Bugfix
- [ ] Feature
- [ ] Code style update
- [ ] Refactor
- [ ] Docs
- [ ] Build-related changes
- [ ] Other, please describe:

If changing the UI of web configuration, please provide the **before/after** screenshot:

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

If yes, please describe the impact and migration path for existing setups:

**The PR fulfills these requirements:**

- [ ] When resolving a specific issue, it's referenced in the PR's body (e.g. `Fixes: #xxx[,#xxx]`, where "xxx" is the issue number)

If adding a **new feature**, the PR's description includes:

- [ ] A convincing reason for adding this feature
- [ ] Related documents have been updated (docs/docs/en)
- [ ] Related tests have been updated

To avoid wasting your time, it's best to open a **feature request issue** first and wait for approval before working on it.

**Other information:**

### CODE REVIEW

File: libsrc/hyperion/LinearColorSmoothing.cpp [hunk 1]
```diff
@@ -84,28 +84,19 @@ void LinearColorSmoothing::handleSettingsUpdate(const settings::type& type, cons
...
 
 		//Debug( _log, "Start Smoothing timer: settlingTime: %d ms, interval: %d ms (%u Hz), updateDelay: %u frames", _settlingTime, _updateInterval, unsigned(1000.0/_updateInterval), _outputDelay );
 		QMetaObject::invokeMethod(_timer, "start", Qt::QueuedConnection, Q_ARG(int, _updateInterval));
 	}
-	else
-	{
-		//std::cout << "LinearColorSmoothing::write> "; LedDevice::printLedValues ( ledValues );
-
-		_targetTime = QDateTime::currentMSecsSinceEpoch() + _settlingTime;
-		memcpy(_targetValues.data(), ledValues.data(), ledValues.size() * sizeof(ColorRgb));
```
Comment (m-seker) [inline libsrc/hyperion/LinearColorSmoothing.cpp:105]: The reason for choosing ````memcpy```` was probably performance concern but most compilers are clever enough to use ````memove```` if the type is POD.

Comment (Lord-Grey) [reply libsrc/hyperion/LinearColorSmoothing.cpp:105]: Given that the smoothing function is part of the critical processing path, I would follow the conclusion that the motivation may had been to reduce processing overhead as much as possible. Some of the hyperion users still use rpi 1 / Zero Hardware...
But I would agree to you that normally Qt copies are  implemented efficiently.

Comment (m-seker) [reply libsrc/hyperion/LinearColorSmoothing.cpp:105]: This one should also have the vectorization possibility, which is not the case for memcpy.

Comment (Lord-Grey) [reply libsrc/hyperion/LinearColorSmoothing.cpp:105]: Sure. Do not misunderstand my comment, I am in support of the proposed change. Just wanted to highlight that not powerful hardware requires still to be considered:).

**Comment by tihoangyeudau** [comment]
I tested this commit and it make my hyperion grabbered incorrrectly. 

I tested on windows.

**Comment by m-seker** [comment]
> I tested this commit and it make my hyperion grabbered incorrrectly.
> 
> I tested on windows and rpi zero

Can you give more details ? What do you observe ?

**Comment by tihoangyeudau** [comment]
> > I tested this commit and it make my hyperion grabbered incorrrectly.
> > I tested on windows and rpi zero
> 
> Can you give more details ? What do you observe ?

sorry your commit is correct. the cause is error of wemos d1 mini device. I change other one then yours was work correctly.

**Comment by Paulchen-Panther** [comment]
Good work. Thank you for your time and effort. 👍

**Pull request merged**